pymodbus/version.py

```diff
@@ -37,7 +37,7 @@ def __str__(self) -> str:
         return f"[{self.package}, version {self.short()}]"
 
 
-version = Version("pymodbus", 3, 1, "x", "")
+version = Version("pymodbus", 3, 1, 4, "alpha")
```

Comment:
@janiversen I think this is better than using `3, 1, "x",` which does not work with `pip`
```
     raise InvalidVersion(f"Invalid version: '{version}'")
pkg_resources.extern.packaging.version.InvalidVersion: Invalid version: '3.1.x'
```

Reply:
This seems very unrelated. I have no problem does this, but it will be after next release.

Reply:
and it does install without problems, at least on my local machine.

Reply:
> This seems very unrelated. I have no problem does this, but it will be after next release.

The only reason for this now was so @jaskij has a simpler way to install and test.

> and it does install without problems, at least on my local machine.

with which command?  `python3 -m pip install -e .` fails for me.  So does `python3 -m pip install --upgrade https://github.com/pymodbus-dev/pymodbus/archive/issue-1276.zip` (until I changed `Version`).

Reply:
I just ran it on my local machine:
```
(pymodbus) pymodbus % pip install -e .
Obtaining file:///Users/jan/repos/pymodbus
  Preparing metadata (setup.py) ... done
Installing collected packages: pymodbus
  Attempting uninstall: pymodbus
    Found existing installation: pymodbus 3.1.3
    Uninstalling pymodbus-3.1.3:
      Successfully uninstalled pymodbus-3.1.3
  Running setup.py develop for pymodbus
Successfully installed pymodbus-3.1.x

[notice] A new release of pip is available: 23.0 -> 23.0.1
[notice] To update, run: pip install --upgrade pip
(pymodbus) pymodbus % 
```
You probably did not run the pip command in your clone.

Your zip file is not a correct installation, it misses important things like wheel, and we should not recommend such installations, people who want to install in between releases need to clone the repository.

Reply:
You must be running `setuptools <= 65.0.0`.  I think this will break for you in `66.0.0`. 
https://github.com/pypa/setuptools/issues/3772

> Your zip file is not a correct installation, it misses important things like wheel, and we should not recommend such installations, people who want to install in between releases need to clone the repository.

I would not recommend for normal use, but it works perfectly fine here, including creating the wheel:
```
python3 -m pip install --upgrade https://github.com/pymodbus-dev/pymodbus/archive/issue-1276.zip
Collecting https://github.com/pymodbus-dev/pymodbus/archive/issue-1276.zip
  Downloading https://github.com/pymodbus-dev/pymodbus/archive/issue-1276.zip
     | 1.1 MB 5.1 MB/s 0:00:00
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: pymodbus
  Building wheel for pymodbus (setup.py) ... done
  Created wheel for pymodbus: filename=pymodbus-3.1.4a0-py2.py3-none-any.whl size=161228 sha256=b29508c6c29a81f43c57d69aede4fdf181f8fb4706123920f7bc378ce76585ac
  Stored in directory: /Users/a.ruddick/Library/Caches/pip/wheels/8e/f6/cf/a802e8a32b1bb7f9fbd8b57bd82eb4cefce5f90d773751cc75
Successfully built pymodbus
Installing collected packages: pymodbus
  Attempting uninstall: pymodbus
    Found existing installation: pymodbus 3.1.x
    Uninstalling pymodbus-3.1.x:
      Successfully uninstalled pymodbus-3.1.x
Successfully installed pymodbus-3.1.4a0
```

Reply:
I am running setuptools 65.4.1, but only for pymodbus. I have venv for each project, and setuptools got installed, when I installed pymodbus. 

I actually wonder how setuptools got installed but I expect it is part of the python installation (which again is in the venv for pymodbus).

I did not know that it creates wheel, but I still think it is dangerous to suggest such installations to users. For a couple of reasons:
- your zip is a snapshot, not frozen in time, so we cannot checkout the version to reproduce problems (install, make a commit, install, and both have the same version).
- users that want to be on the bleeding edge should clone the repository, what allows them to get the latest commits easy.

apart from that, I noticed that "pip install -e ." also works with a standard python 3.11 installation.

I have however no problem in you recommending what you think is right,, but I do have a problem changing version on dev without a release. After a release we can (as I wrote earlier) decide if we want to change the scheme.

Reply:
https://github.com/pymodbus-dev/pymodbus/issues/1372